File: src/main/java/utils/OpticsList.java
package utils;

import optics.objects.Interactive;

import java.util.ArrayList;
import java.util.Collections;

public class OpticsList<T extends Interactive> extends ArrayList<T> {

  public <T> OpticsList() {
    super();
  }

  public void addAll(T... args){
    Collections.addAll(this, args);
  }

  public OpticsList<T> getAllExcept(T exclude){
    OpticsList<T> result = new OpticsList<>();
    for(T elem: this){
      if(!elem.equals(exclude))result.add(elem);
    }
    return result;
  }

  public OpticsList<T> deepClone(){
    OpticsList<T> result = new OpticsList<>();
    for(T elem: this){
      result.add((T) elem.cloneObject());
    }
    return result;
  }
}

File: src/main/java/utils/OpticsList.kt
package utils

import optics.objects.Interactive
import java.util.*

class OpticsList<T : Interactive<T>> : ArrayList<T>() {

    fun getAllExcept(exclude: T): OpticsList<T> {
        val result: OpticsList<T> = OpticsList()
        for (elem in this) {
            if (elem != exclude) result.add(elem)
        }
        return result
    }

    fun deepClone(): OpticsList<T> {
        val result: OpticsList<T> = OpticsList()
        for (elem in this) {
            result.add(elem.cloneObject())
        }
        return result
    }
}
